src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardSyncTaskManager.java

```diff
@@ -117,30 +117,7 @@
                             cleanupLeasesUponShardCompletion, ignoreUnexpectedChildShards, shardSyncIdleTimeMillis,
                             shardSyncer, latestShards), metricsFactory);
             future = CompletableFuture.supplyAsync(() -> currentTask.call(), executorService)
-                    .whenComplete((taskResult, exception) -> {
-                        if (exception != null) {
-                            LOG.error("Caught exception running " + currentTask.getTaskType() + " task: ", exception);
-                        }
-                        // Acquire lock here. If shardSyncRequestPending is false in this completionStage and
-                        // syncShardAndLeaseInfo is invoked, before completion stage exits (future completes)
-                        // but right after the value of shardSyncRequestPending is checked, it will result in
-                        // shardSyncRequestPending being set to true, but no pending futures to trigger the next
-                        // ShardSyncTask. By executing this stage in a Reentrant lock, we ensure that if the
-                        // previous task is in this completion stage, checkAndSubmitNextTask is not invoked
-                        // until this completionStage exits.
-                        try {
-                            lock.lock();
-                            if (shardSyncRequestPending.get()) {
-                                shardSyncRequestPending.set(false);
-                                // reset future to null, so next call creates a new one
-                                // without trying to get results from the old future.
-                                future = null;
-                                checkAndSubmitNextTask(latestShards);
```

Comment:
Adding the offline comment given to Parijat, before introducing the lock here, for visibility:

     If shardSyncRequestPending is false in this completionStage and
     syncShardAndLeaseInfo is invoked, before completion stage exits (future completes)
     but right after the value of shardSyncRequestPending is checked, it will result in
     shardSyncRequestPending being set to true, but no pending futures to trigger the next
     ShardSyncTask. We need to execute this in a Reentrant lock, we ensure that if the
     previous task is in this completion stage, checkAndSubmitNextTask is not invoked
     until this completionStage exits.